server/service/osquery.go

```diff
@@ -1405,3 +1411,139 @@ func (svc *Service) SubmitResultLogs(ctx context.Context, logs []json.RawMessage
 	}
 	return nil
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Query Reports
+////////////////////////////////////////////////////////////////////////////////
+
+const (
+	maxQueryReportRows = 1000
+)
+
+func (svc *Service) SaveResultLogsToQueryReports(ctx context.Context, results []json.RawMessage) error {
+	// skipauth: Authorization is currently for user endpoints only.
+	svc.authz.SkipAuthorization(ctx)
+
+	// Do not insert results if query reports are disabled globally
+	appConfig, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return newOsqueryError("getting app config: " + err.Error())
+	}
+	if appConfig.ServerSettings.QueryReportsDisabled {
+		return nil
+	}
+
+	var queryResults []fleet.ScheduledQueryResult
+	for _, raw := range results {
+		var result fleet.ScheduledQueryResult
+		if err := json.Unmarshal(raw, &result); err != nil {
+			return newOsqueryError("unmarshalling individual result log: " + err.Error())
+		}
+		queryResults = append(queryResults, result)
+	}
+
+	// Filter results to only the most recent for each query
+	filtered := getMostRecentResults(queryResults)
+
+	for _, result := range filtered {
+		if err := svc.processResults(ctx, result); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (svc *Service) processResults(ctx context.Context, result fleet.ScheduledQueryResult) error {
+	query, err := svc.ds.QueryByName(ctx, nil, getQueryNameFromResult(result.QueryName))
```

Comment:
Let's parse the teamID from the scheduled query name:
```go
queryName, teamID := getQueryNameFromResult(result.QueryName)
```

- pack/team-$ID/$queryName for queries that belong to teams.
- pack/Global/$queryName for global queries.